pyteal/ast/frame.py

```diff
@@ -0,0 +1,192 @@
+from pyteal.errors import TealInputError, verifyProgramVersion
+from pyteal.ir import TealBlock, TealSimpleBlock, TealOp, Op
+
+if TYPE_CHECKING:
+    from pyteal.compiler import CompileOptions
+
+
+class Proto(Expr):
+    def __init__(self, num_args: int, num_returns: int):
+        super().__init__()
+        if num_args < 0:
+            raise TealInputError(
+                f"the number of arguments provided to Proto must be >= 0 but {num_args=}"
+            )
+        if num_returns < 0:
+            raise TealInputError(
+                f"the number of return values provided to Proto must be >= 0 but {num_returns=}"
+            )
+        self.num_args = num_args
+        self.num_returns = num_returns
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        verifyProgramVersion(
+            Op.proto.min_version,
+            options.version,
+            "Program version too low to use op proto",
+        )
+        op = TealOp(self, Op.proto, self.num_args, self.num_returns)
+        return TealBlock.FromOp(options, op)
+
+    def __str__(self) -> str:
+        return f"(proto: num_args = {self.num_args}, num_returns = {self.num_returns})"
+
+    def type_of(self) -> TealType:
+        return TealType.none
+
+    def has_return(self) -> bool:
+        return False
+
+
+Proto.__module__ = "pyteal"
+
+
+class FrameDig(Expr):
+    def __init__(self, depth: int):
+        super().__init__()
+        self.depth = depth
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        verifyProgramVersion(
+            Op.frame_dig.min_version,
+            options.version,
+            "Program version too low to use op frame_dig",
+        )
+        op = TealOp(self, Op.frame_dig, self.depth)
+        return TealBlock.FromOp(options, op)
+
+    def __str__(self) -> str:
+        return f"(frame_dig: dig_depth = {self.depth})"
+
+    def type_of(self) -> TealType:
+        return TealType.anytype
+
+    def has_return(self) -> bool:
+        return False
+
+
+FrameDig.__module__ = "pyteal"
+
+
+class FrameBury(Expr):
+    def __init__(self, what: Expr, depth: int):
+        super().__init__()
+        require_type(what, TealType.anytype)
+        self.what = what
+        self.depth = depth
+
+    def __teal__(self, options: "CompileOptions") -> tuple[TealBlock, TealSimpleBlock]:
+        verifyProgramVersion(
+            Op.frame_bury.min_version,
+            options.version,
+            "Program version too low to use op frame_bury",
+        )
+        op = TealOp(self, Op.frame_bury, self.depth)
+        return TealBlock.FromOp(options, op, self.what)
+
+    def __str__(self) -> str:
+        return f"(frame_bury (bury_depth = {self.depth}) ({self.what}))"
+
+    def type_of(self) -> TealType:
+        return TealType.none
+
+    def has_return(self) -> bool:
+        return False
+
+
+FrameBury.__module__ = "pyteal"
+
+
+class Bury(Expr):
```

Comment:
Should we take this as an opportunity to expose `Dig` as well? I don't have a strong opinion here, but it's worth a quick convo.

Reply:
I see two (or three) options here:
- expose `Dig` for currently exposed `Bury`
- eliminate `Bury` for `Dig` is not currently exposed here
- do nothing about `Dig`

I prefer first 2, for I preferring providing all or none, rather than giving something in the middle.

Currently I feel hard to use `Dig/Bury` in code, for it strictly require one knowing what is there some `depth` down the stack. But exposing both opcode does not seem a harm to me.

Reply:
@michaeldiamant @jasonpaulos - Any thoughts? Do we see any immediate use cases for `Bury`? And would those use cases not also necessitate providing `Dig`?

If not, I propose eliminating `Bury` from the PR.

Reply:
I don't see any reason to expose `Dig` and/or `Bury` to users. Directly stack manipulation is too low-level for PyTeal.

Reply:
got it, I am opt to take that out